Sporadic error `call on null client` for first RPC call after upgrading from 3.0.0-alpha.30 to 3.0.1-alpha.1, ~1% chance

After updating to 3.0.1-alpha.1 from 3.0.0-alpha.30, a load test of mine now fails occasionally with error `schema/avas/corehost/corehost.capnp:CoreFactory.buildInfo: call on null client`.

__I am trying to build a minimal reproducer to narrow this down and post it here.__ (But that might take some time, I'm deferring to update for now.) In the meantime I'd leave the following details here in case someone else stumbles over the issue:

- The environment is a Go HTTP server, which connects to a C++ capnp RPC server within a HTTP request.
- 0.1-2% of requests fail with capnp 3.0.1-alpha.1 regardless for request rate or parallel requests (high variance in failure rate). More requests make it fail more often, linear correlation.
- The affected code is heavily used in production with tens of thousands of daily connections, without issues with capnp 3.0.0-alpha.30.
- No errors are logged to `Logger`.

The affected code is as simple as the following after a successful connection:
```
// (establishment + error checks for tcpConn omitted)
conn := rpc.NewConn(rpc.NewStreamTransport(tcpConn), &rpc.Options{
	Logger: &rpcErrorReporter{sess: s},
})
factory := corehost.CoreFactory(conn.Bootstrap(s.ctx))
buildInfoFuture, buildInfoRelease := factory.BuildInfo(s.ctx, nil)
defer buildInfoRelease()
buildInfo, err := buildInfoFuture.Struct() // err is not nil for 0.2-2% of requests
```

Running tests on main nets me the following:
```
--- FAIL: TestPromiseOrdering (0.08s)
    senderpromise_test.go:404:
        	Error Trace:	/Users/okuckertz/Devel/go-capnp/rpc/senderpromise_test.go:404
        	Error:      	Received unexpected error:
        	            	test.capnp:PingPong.echoNum: call on null client
        	Test:       	TestPromiseOrdering
        	Messages:   	call #0 should succeed
FAIL
FAIL	capnproto.org/go/capnp/v3/rpc	0.550s
```

So this is probably related to or a duplicate of #543.

@mologie Did you still got this error:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xb6da5e]

goroutine 2416337 [running]:
capnproto.org/go/capnp/v3/rpc.(*Conn).handleCall.func4.6()
	capnproto.org/go/capnp/v3@v3.0.0-alpha.30/rpc/rpc.go:952 +0x3e
zenhack.net/go/util/deferred.(*Queue).Run(0x1218cf0?)
	zenhack.net/go/util@v0.0.0-20230414204917-531d38494cf5/deferred/queue.go:31 +0x42
capnproto.org/go/capnp/v3/rpc.(*Conn).handleCall(0xc022946b00, {0x12251c0, 0xc020f96c80}, {0x12210e8, 0xc0212b0420})
	capnproto.org/go/capnp/v3@v3.0.0-alpha.30/rpc/rpc.go:844 +0xaba
capnproto.org/go/capnp/v3/rpc.(*Conn).receive.func1()
	capnproto.org/go/capnp/v3@v3.0.0-alpha.30/rpc/rpc.go:604 +0x325
capnproto.org/go/capnp/v3/rpc.(*Conn).receive.(*Conn).backgroundTask.func2()
	capnproto.org/go/capnp/v3@v3.0.0-alpha.30/rpc/rpc.go:286 +0x62
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.10.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 2416334
	golang.org/x/sync@v0.10.0/errgroup/errgroup.go:75 +0x96
# Errors occurred during execution: exit status 2
```

I tried to use `v3.0.0-alpha.30`, but this panic still happened, but with a much lower probability (may be 1 in 10,000,000 reqs)

--- a/rpc/level0_test.go
+++ b/rpc/level0_test.go
@@ -8,6 +8,7 @@ import (
 	"net"
 	"os"
 	"runtime"
+	"strconv"
 	"strings"
 	"sync"
 	"testing"
@@ -1246,6 +1247,22 @@ func TestRecvBootstrapCallException(t *testing.T) {
 // requirement.
 func TestRecvBootstrapPipelineCall(t *testing.T) {
 	t.Parallel()
+	testRecvBootstrapPipelineCall(t)
+}
+
+func TestRecvBootstrapPipelineCallStress(t *testing.T) {
+	if os.Getenv("CAPNP_RPC_STRESS") != "1" {
+		t.Skip("set CAPNP_RPC_STRESS=1 to run promised-answer lifecycle stress test")
+	}
+
+	for i := 0; i < 100; i++ {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			testRecvBootstrapPipelineCall(t)
+		})
+	}
+}
+
+func testRecvBootstrapPipelineCall(t *testing.T) {
 
 	srvShutdown := make(chan struct{})
 	srv := newServer(

--- a/rpc/level0_test.go
+++ b/rpc/level0_test.go
@@ -1250,6 +1250,40 @@ func TestRecvBootstrapPipelineCall(t *testing.T) {
 	testRecvBootstrapPipelineCall(t)
 }
 
+func TestRecvPromisedAnswerMissingCapability(t *testing.T) {
+	t.Parallel()
+	srv := newServer(func(_ context.Context, call *server.Call) error {
+		results, err := call.AllocResults(capnp.ObjectSize{PointerCount: 1})
+		if err != nil {
+			return err
+		}
+		results.SetPtr(0, capnp.NewInterface(results.Segment(), 1).ToPtr())
+		return nil
+	}, nil)
+	left, right := transport.NewPipe(1)
+	conn := rpc.NewConn(rpc.NewTransport(left), &rpc.Options{BootstrapClient: srv, Logger: testErrorReporter{tb: t}})
+	p := rpc.NewTransport(right)
+	defer finishTest(t, conn, p)
+	ctx := context.Background()
+	const boot, source, bad, good = 1, 2, 3, 4
+	require.NoError(t, sendMessage(ctx, p, &rpcMessage{Which: rpccp.Message_Which_bootstrap, Bootstrap: &rpcBootstrap{QuestionID: boot}}))
+	_, err := recvBootstrapReturn(ctx, p, boot)
+	require.NoError(t, err)
+	send := func(qid uint32, target rpcMessageTarget) *rpcMessage {
+		require.NoError(t, sendMessage(ctx, p, &rpcMessage{Which: rpccp.Message_Which_call, Call: &rpcCall{QuestionID: qid, Target: target, InterfaceID: interfaceID, MethodID: methodID}}))
+		m, release, err := recvMessage(ctx, p)
+		require.NoError(t, err)
+		defer release()
+		return m
+	}
+	first := send(source, rpcMessageTarget{Which: rpccp.MessageTarget_Which_promisedAnswer, PromisedAnswer: &rpcPromisedAnswer{QuestionID: boot}})
+	require.Equal(t, rpccp.Return_Which_results, first.Return.Which)
+	badReturn := send(bad, rpcMessageTarget{Which: rpccp.MessageTarget_Which_promisedAnswer, PromisedAnswer: &rpcPromisedAnswer{QuestionID: source, Transform: []rpcPromisedAnswerOp{{Which: rpccp.PromisedAnswer_Op_Which_getPointerField, GetPointerField: 0}}}})
+	require.Equal(t, rpccp.Return_Which_exception, badReturn.Return.Which)
+	valid := send(good, rpcMessageTarget{Which: rpccp.MessageTarget_Which_promisedAnswer, PromisedAnswer: &rpcPromisedAnswer{QuestionID: boot}})
+	require.Equal(t, rpccp.Return_Which_results, valid.Return.Which)
+}
+
 func TestRecvBootstrapPipelineCallStress(t *testing.T) {
 	if os.Getenv("CAPNP_RPC_STRESS") != "1" {
 		t.Skip("set CAPNP_RPC_STRESS=1 to run promised-answer lifecycle stress test")

--- a/rpc/promised_target_test.go
+++ b/rpc/promised_target_test.go
@@ -1,0 +1,50 @@
+package rpc
+
+import (
+	"errors"
+	"testing"
+
+	"capnproto.org/go/capnp/v3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPromisedAnswerTarget(t *testing.T) {
+	t.Parallel()
+
+	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
+	require.NoError(t, err)
+	defer msg.Release()
+
+	t.Run("MissingCapability", func(t *testing.T) {
+		target, err := promisedAnswerTarget(capnp.NewInterface(seg, 1).ToPtr(), nil)
+		assert.Error(t, err)
+		assert.False(t, target.IsValid())
+	})
+
+	t.Run("InvalidCapability", func(t *testing.T) {
+		target, err := promisedAnswerTarget(capnp.NewInterface(seg, 0).ToPtr(), []capnp.ClientSnapshot{{}})
+		assert.Error(t, err)
+		assert.False(t, target.IsValid())
+	})
+
+	t.Run("NonCapability", func(t *testing.T) {
+		st, err := capnp.NewStruct(seg, capnp.ObjectSize{})
+		require.NoError(t, err)
+		target, err := promisedAnswerTarget(st.ToPtr(), nil)
+		require.NoError(t, err)
+		defer target.Release()
+		assert.True(t, target.IsValid())
+	})
+
+	t.Run("Capability", func(t *testing.T) {
+		client := capnp.ErrorClient(errors.New("test"))
+		defer client.Release()
+		snapshot := client.Snapshot()
+		defer snapshot.Release()
+		target, err := promisedAnswerTarget(capnp.NewInterface(seg, 0).ToPtr(), []capnp.ClientSnapshot{snapshot})
+		require.NoError(t, err)
+		defer target.Release()
+		assert.True(t, target.IsValid())
+	})
+}

--- a/rpc/rpc.go
+++ b/rpc/rpc.go
@@ -927,16 +927,11 @@ func (c *Conn) handleCall(ctx context.Context, in transport.IncomingMessage) err
 					dq.Defer(in.Release)
 					return nil
 				}
-				iface := sub.Interface()
-				var tgt capnp.ClientSnapshot
-				if sub.IsValid() && !iface.IsValid() {
-					tgt = capnp.ErrorClient(rpcerr.Failed(ErrNotACapability)).Snapshot()
-				} else {
-					capID := iface.Capability()
-					capTable := tgtAns.returner.resultsCapTable
-					if int(capID) < len(capTable) {
-						tgt = capTable[capID].AddRef()
-					}
+				tgt, err := promisedAnswerTarget(sub, tgtAns.returner.resultsCapTable)
+				if err != nil {
+					ans.sendException(dq, err)
+					dq.Defer(in.Release)
+					return nil
 				}
 
 				c.tasks.Add(1) // will be finished by answer.Return
@@ -967,6 +962,26 @@ func (c *Conn) handleCall(ctx context.Context, in transport.IncomingMessage) err
 			panic("unreachable")
 		}
 	})
+}
+
+// promisedAnswerTarget resolves an interface pointer from a promised answer's
+// results. The returned snapshot owns a reference and must be released by the
+// caller. A non-capability result is represented by an error client; a missing
+// capability-table entry is an invalid incoming call.
+func promisedAnswerTarget(sub capnp.Ptr, capTable []capnp.ClientSnapshot) (capnp.ClientSnapshot, error) {
+	iface := sub.Interface()
+	if !sub.IsValid() || !iface.IsValid() {
+		return capnp.ErrorClient(rpcerr.Failed(ErrNotACapability)).Snapshot(), nil
+	}
+
+	capID := iface.Capability()
+	if int(capID) >= len(capTable) || !capTable[capID].IsValid() {
+		return capnp.ClientSnapshot{}, rpcerr.Failed(errors.New(
+			"incoming call: promised answer capability index " + str.Utod(capID) +
+				" is not present in results capability table",
+		))
+	}
+	return capTable[capID].AddRef(), nil
 }
 
 // A promisedPipelineCaller is a PipelineCaller that stands in for another

